src/event.rs

```diff
@@ -21,19 +25,39 @@ pub enum Event<'a> {
     UncaughtError(&'a Error),
 }
 
-pub(crate) struct EventListener {
-    f: Box<dyn Fn(Event) + Send + Sync + 'static>,
+pub(crate) struct EventListener<State> {
+    callback: EventCallback<State>,
 }
 
-impl EventListener {
-    pub fn new<F>(f: F) -> Self
-    where
-        F: Fn(Event) + Send + Sync + 'static,
-    {
-        Self { f: Box::new(f) }
+impl<State> EventListener<State> {
+    pub fn new(callback: EventCallback<State>) -> Self {
+        Self { callback }
+    }
+
+    pub fn call(&self, event: Event, state: &Arc<State>) {
+        // TODO:
+        //
+        // Consider implementing an integrity check of some kind before calling
+        // the callback. If the performance overhead is acceptable and doing so
+        // mitigates the risk of an attacker exploiting `callback`, we'll avoid
+        // having to clone Arc<EventListener> twice for every request.
+        //
+        // Alternatives:
+        // - Use Arc<dyn Fn(Event, &Arc<State>) + Copy + Send + Sync + 'static>
+        // - Use a channel to send events to a stream that exists for the
+        //   lifetime of the application and spawn the event loop that accepts
+        //   incoming TCP connections in a separate tokio task.
+        //
+        (self.callback)(event, state)
```

Comment:
I'm considering refactoring the `App::listen` to return a stream of events as a safer approach to passing around a function pointer and using an `mpsc::channel`. I'm less concerned about the performance of an `Arc::clone` or cloning a `sender` as much as I'm concerned about giving users a safe and idiomatic way of responding the events that occur outside the context of a request / response cycle. While the errors that are bubbled up to the event listener are not critical application errors, knowing of their existence is a game changer when you are tasked with fixing strange or unexpected behavior in an application.